server/src/main/java/com/thoughtworks/go/server/service/AgentService.java

```diff
@@ -126,143 +140,409 @@ public AgentsViewModel registeredAgents() {
+        if (agent.isNull()) {
+            String agentNotFoundMessage = format("Agent '%s' not found", agent.getUuid());
+            result.notFound("Agent not found.", agentNotFoundMessage, general(GLOBAL));
             return true;
         }
+
         return false;
     }
 
-    private boolean hasOperatePermission(Username username, OperationResult operationResult) {
+    private boolean doesNotHaveOperatePermission(Username username, OperationResult result) {
         if (!securityService.hasOperatePermissionForAgents(username)) {
             String message = "Unauthorized to operate on agent";
-            operationResult.forbidden(message, message, HealthStateType.general(HealthStateScope.GLOBAL));
-            return false;
+            result.forbidden(message, message, general(GLOBAL));
+            return true;
         }
-        return true;
+        return false;
     }
 
-    public AgentInstance updateAgentAttributes(Username username, HttpOperationResult result, String uuid, String newHostname, String resources, String environments, TriState enable) {
-        if (!hasOperatePermission(username, result)) {
+    public AgentInstance updateAgentAttributes(Username username, HttpOperationResult result, String uuid,
+                                               String hostname, String resources, EnvironmentsConfig environments,
+                                               TriState state) {
+        AgentUpdateValidator validator = new AgentUpdateValidator(username, agentInstances.findAgent(uuid), hostname, environments,
+                resources, state, result, goConfigService);
+        Agent agent = agentDao.agentByUuid(uuid);
+        try {
+            if (validator.canContinue()) {
+                validator.validate();
+
+                setAgentAttributes(hostname, resources, state, agent);
+                agent.setEnvironments("");
+                addOnlyThoseEnvsThatAreNotAssociatedWithTheAgentFromConfigRepo(environments, agent);
+                saveOrUpdate(agent);
+
+                if (agent.hasErrors()) {
+                    result.unprocessibleEntity("Updating agent failed.", "", general(GLOBAL));
+                } else {
+                    result.ok(format("Updated agent with uuid %s.", agent.getUuid()));
+                }
+            }
+        } catch (InvalidPendingAgentOperationException | IllegalArgumentException | RecordNotFoundException e) {
+            LOGGER.error(e.getMessage(), e);
+            return null;
+        } catch (Exception e) {
+            String msg = "Updating agent failed: " + e.getMessage();
+            LOGGER.error(msg, e);
+            result.internalServerError(msg, general(GLOBAL));
             return null;
         }
 
-        AgentInstance agentInstance = findAgent(uuid);
-        if (isUnknownAgent(agentInstance, result)) {
-            return null;
+        return AgentInstance.createFromAgent(agent, systemEnvironment, agentStatusChangeNotifier);
```

Comment:
This comment is optional.
I think it is redundant to create a AgentInstance here, earlier the AgentConfig did not have all the attributes of an agent guess that's the reason for creating one. Now since the Agent object has all the attributes, the same can be passed back.

Reply:
This method is called from `AgentsControllerV5` and while creating json response to return, it requires fields that are not present in `Agent` object. Some of those fields are runtime information which is available only in `AgentInstance`.

Currently,
- `Agent` - represents a record in Agents table
- `AgentInstance` - contains runtime information about an agent. That includes,
     - Type of an agent (Local or remote)
     - Status (Idle, Building, LostContact, Missing, Cancelled, Unknown)
     - Build Info (Names of the pipeline, stage, job, and whether it is currently building or not)
     - Agent record in DB (Agent class object)
     - And many more

Reply:
Fair enough!